### ISSUE TITLE
Support uv

### ISSUE THREAD

**Issue by njzjz** [readthedocs/readthedocs.org#11289]
## What's the problem this feature will solve?

Speed up Python installation using [uv](https://github.com/astral-sh/uv), an extremely fast Python package installer that can replace pip. I have seen uv faster than pip in all of my Python projects.

## Describe the solution you'd like

Provide an option to replace `python -m pip` below with `python -m uv pip`.

![image](https://github.com/readthedocs/readthedocs.org/assets/9496702/a351de5e-a015-48ef-a57b-7a0dac2e8289)

I think the related code is here:

https://github.com/readthedocs/readthedocs.org/blob/791643f452b872fc985f158a7051f38a65f5f6ce/readthedocs/doc_builder/python_environments.py#L150-L157

https://github.com/readthedocs/readthedocs.org/blob/791643f452b872fc985f158a7051f38a65f5f6ce/readthedocs/doc_builder/python_environments.py#L201-L206

and uv is required to be installed in advance.

## Alternative solutions

None

## Additional context

- uv project: https://github.com/astral-sh/uv

**Comment by humitos** [comment]
Hi! We currently don't support `uv` at the core level and we are not planning to change the default `pip` behavior in the near future. _However_, users wanting to use `uv` can build their project using `build.jobs.post_install` (similar to how people is using `poetry`: https://docs.readthedocs.io/en/latest/build-customization.html#install-dependencies-with-poetry)

On the other hand, the whole build process can be overwritten by using `build.commands` (https://docs.readthedocs.io/en/latest/build-customization.html) and running all the commands manually, including `uv` to install the requirements.

Please, let me know in a following comment if you were able to make it work with any of these approaches. That way, users arriving to this issue will find an immediately solution.

Finally, it's worth to mention that we have plans to support overriding pre-defined jobs like `build.jobs.install` and `build.jobs.build` in the following quarter (hopefully) [^1]. With that in place, users will be able to do something like the following to override `pip install` command:

```yaml
version: 2
build:
  jobs:
    install:
      - uv pip sync docs/requirements.txt
```

[^1]: internal and private conversation about this: https://github.com/readthedocs/meta/discussions/46. We will eventually make this task a public issue to receive feedback from our users.

**Comment by humitos** [comment]
Great, thanks! I opened a PR with a small documentation section showing how to use `uv` currently: https://github.com/readthedocs/readthedocs.org/pull/11290

**Comment by henryiii** [comment]
By the way, the two remaining steps are really slow: `python -m pip install --upgrade --no-cache-dir pip setuptools` and `python -m pip install --upgrade --no-cache-dir sphinx readthedocs-sphinx-ext`. If there was some way to override these to be `uv pip` instead, I expect a lot of that remaining time would be gone. They seem to take 40 seconds; while uv should be able to do it in a few seconds, probably more than halving your remaining time in the example above.

**Comment by henryiii** [comment]
The following workaround does work:
```yaml
build:
  jobs:
    post_create_environment:
      - pip install uv
      - VIRTUAL_ENV=$READTHEDOCS_VIRTUALENV_PATH uv pip install --upgrade pip setuptools
      - VIRTUAL_ENV=$READTHEDOCS_VIRTUALENV_PATH uv pip install --upgrade sphinx readthedocs-sphinx-ext
    post_install:
      - VIRTUAL_ENV=$READTHEDOCS_VIRTUALENV_PATH uv pip install .[docs]
```

UV resolves the first line in 0 seconds (not measurable) vs. pip taking 18 seconds. And the second takes 1 seconds (down from 20). Then pip wastes two seconds checking and reporting that everything is up to date. So this changes 40 seconds into <4 seconds. ;)

**Comment by henryiii** [comment]
You can drop the uv install time to 4 seconds with curl, but the fastest (and least annoying way) is to use asdf, there's a plugin: https://github.com/asdf-community/asdf-uv. That gets it in 0s + 0s + 1s, which regardless of rounding should be less than four seconds, and doesn't require adding a bunch of full paths to `~/.cargo/bin/uv`. But it still wastes time on the pip steps that aren't doing anything (and it's installing things that aren't needed, like pip and setuptools). The fastest method is this:

```python
build:
  os: "ubuntu-22.04"
  tools:
    python: "3.12"
  commands:
    - asdf plugin add uv
    - asdf install uv latest
    - asdf global uv latest
    - uv venv
    - uv pip install .[docs]
    - .venv/bin/python -m sphinx -T -b html -d docs/_build/doctrees -D language=en docs $READTHEDOCS_OUTPUT/html
```

The entire build takes 21 seconds, less than half of the time it takes with the default pip configuration.

(Would adding `tools: uv: "latest"` be something easy and reasonable to do?)

**Comment by humitos** [comment]
Thanks @henryiii for all the research on this and for sharing it on this issue. I think it will help other users having the same doubts. We could probably update our documentation to reflect what you posted using `build.commands` 👍🏼 

> (Would adding `tools: uv: "latest"` be something easy and reasonable to do?)

Considering that installing `uv` via `asdf` as you mentioned only takes 1s, we won't earn too much time by adding it as an option for `build.tools` (they live in S3 and are copied at build time into the Docker image).

I would say that the `build.commands` example that you posted here is the way to go for users that want to use `uv` to speed up their builds ⚡

**Comment by ddelange** [comment]
Hi 👋 We currently have the following config:

```yaml
---
# .readthedocs.yml
# Read the Docs configuration file
# See https://docs.readthedocs.io/en/stable/config-file/v2.html for details

# Required
version: 2

# Set the OS, Python version and other tools you might need
build:
  os: ubuntu-22.04
  tools:
    python: '3.12'

# Build documentation in the docs/ directory with Sphinx
sphinx:
  configuration: docs/source/conf.py

# Build documentation with MkDocs
#mkdocs:
#  configuration: mkdocs.yml

# Optionally build your docs in additional formats such as PDF and ePub
formats: all

# Optional but recommended, declare the Python requirements required
# to build your documentation
# See https://docs.readthedocs.io/en/stable/guides/reproducible-builds.html
python:
  install:
  - method: pip
    path: .
  - requirements: requirements_rtd.txt
```

It would be cool if we can change `- method: pip` to `- method: uv`, instead of adding a whole new `commands` section as described above. Is that worth re-opening this issue?

**Comment by EwoutH** [comment]
I would love to have improved `uv` support in Readthedocs, making it as simple defining:

```yaml
python:
  install:
    - method: uv
```

Would you like me to open a new issue for that or shall we reopen this issue and discuss it here further?

**Comment by ddelange** [comment]
please see #11551 for a proposal that would add more options for the user than just `uv`

**Comment by ddelange** [comment]
not saying I'm against a `method: uv` alternative, feels like a simpler solution:)

**Comment by humitos** [comment]
As @ddelange mentioned, we are going to move forward with a more _generic_ solution here. Implementing `python.install.method: uv` will require a lot of effort from the core team and will solve only a particular and specific use case. We want to support more tools (e.g. poetry, pdm, etc) and we also want to provide users with a more customizable experience like passing extra arguments to the these command.

**Comment by ddelange** [comment]
Another alternative is to commit and maintain a lockfile in your repo using uv, and point RTD to that file for a regular pip install (which hence will be fast).
```sh
uv pip compile requirements/*.txt -o requirements/requirements.lock --no-annotate --no-header
```

Example enforcement in Github Actions:
```yaml
    - name: Check whether requirements changed
      id: requirements-changed
      uses: tj-actions/changed-files@v34
      with:
        files: |
          requirements/*.txt

    - name: Check whether make lockfile was forgotten
      if: steps.requirements-changed.outputs.any_changed == 'true'
      run: |
        uv pip compile requirements/*.txt -o requirements/requirements.lock --no-annotate --no-header
        test ! "$(git diff)" && echo "no changes" || ( git diff && exit 1 )
```

**Comment by humitos** [comment]
We've deployed some changes that allow people to use `uv` in a simpler way. Here is an example of the YAML used to build a Sphinx project using `uv` to install all the dependencies.

```yaml
version: 2

build:
  os: ubuntu-24.04
  tools:
    python: "3.13"
  jobs:
    pre_create_environment:
      - asdf plugin add uv
      - asdf install uv latest
      - asdf global uv latest
    create_environment:
      - uv venv
    install:
      - uv pip install -r requirements.txt
    build:
      html:
        - uv run sphinx-build -T -b html docs $READTHEDOCS_OUTPUT/html
```

- Read the Docs build: https://app.readthedocs.org/projects/test-builds/builds/26775264/
- Built documentation: https://test-builds.readthedocs.io/en/uv/

Let us know if this works for your use case.

**Comment by EwoutH** [comment]
@humitos thanks, that looks interesting. We're currently not using the `jobs` tag in [our configuration](https://github.com/projectmesa/mesa/blob/main/.readthedocs.yml), can we also define it in `python.install`? For example by changing the method to `uv`, `uv pip` or something like it?

```yml
version: 2

# Build documentation in the docs/ directory with Sphinx
sphinx:
  configuration: docs/conf.py

build:
  os: ubuntu-lts-latest
  tools:
    python: latest

# Optionally set the version of Python and requirements required to build your docs
python:
  install:
    - method: pip
      path: .
      extra_requirements:
        - docs

```

**Comment by humitos** [comment]
> can we also define it in `python.install`? For example by changing the method to `uv`, `uv pip` or something like it?

No, it's not possible. To use `uv` you need to use `build.jobs.*`.

**Comment by effigies** [comment]
Updating to use the new `create_environment` and `install` stages, my [previous config](https://github.com/readthedocs/readthedocs.org/issues/11289#issuecomment-2465248908) reduces to:

```yaml
version: 2

build:
  os: ubuntu-lts-latest
  tools:
    python: latest
  jobs:
    pre_create_environment:
      - asdf plugin add uv
      - asdf install uv latest
      - asdf global uv latest
    create_environment:
      - uv venv $READTHEDOCS_VIRTUALENV_PATH
    install:
      # Use a cache dir in the same mount to halve the install time
      - VIRTUAL_ENV=$READTHEDOCS_VIRTUALENV_PATH uv pip install --cache-dir $READTHEDOCS_VIRTUALENV_PATH/../../uv_cache -r docs/requirements.txt

sphinx:
  configuration: docs/conf.py
```

Using the `READTHEDOCS_VIRTUALENV_PATH` still helps avoid reproducing the build job. Setting `--cache-dir` allows hard links, which don't work by default.

**Comment by henryiii** [comment]
Can this be used with `uv sync` or even `uv  run` directly? `uv pip` doesn’t support dependency-groups (PEP 735) yet (waiting on pip support), but the high level interface does. Would I just have to put a placeholder in the unused stages?

**Comment by stsewd** [comment]
> [@henryiii](https://github.com/henryiii) here's our config using uv sync and uv run.
> 
> https://github.com/aiidateam/aiida-core/blob/main/.readthedocs.yml
> 
> You're correct that you need to overwrite the stages that uv doesn't necessarily need (e.g. we're creating virtual environment with `uv venv` even though uv sync would take care of that automatically)

@danielhollas if you don't include the `sphinx` key, the default steps won't run, so you don't need to override the create environment step if you don't need to.

**Comment by cjw296** [comment]
Time for this to be re-opened? Life is so much easier with uv, it's sad that RTD doesn't offer any support...

**Comment by humitos** [comment]
> Time for this to be re-opened? Life is so much easier with uv, it's sad that RTD doesn't offer any support...

Hace you tried https://docs.readthedocs.com/platform/stable/build-customization.html#install-dependencies-with-uv ?

**Comment by cjw296** [comment]
Yes, I use that hack on my projects:
- https://github.com/cjw296/xerotrust/blob/main/.readthedocs.yaml
- https://github.com/simplistix/paradigms/blob/main/.readthedocs.yaml

Honestly, given how much more performant `uv` is than `pip`, I'd advocate for it becoming the default for RTD...

**Comment by ddelange** [comment]
even if not the entire uv ecosystem, a `uv pip install -r ...` alternative would be such a time saver:
```diff
python:
  install:
- - method: pip
+ - method: uv
    path: .
  - requirements: requirements_rtd.txt
```

**Comment by cjw296** [comment]
@humitos: Also, note that much config is ignored with the hack, this (while required) becomes a misleading lie:

```yaml
  tools:
    python: "3.13"
```

...so I've ended up just using this, which feels very unpleasant:

```yaml
version: 2

build:
  os: ubuntu-24.04
  commands:
  - asdf plugin add uv
  - asdf install uv latest
  - asdf global uv latest
  - uv sync --group docs
  - uv run sphinx-build --fail-on-warning -T -b html docs $READTHEDOCS_OUTPUT/html
```

**Comment by EwoutH** [comment]
With so many projects using Readthedocs, I think there’s also significant environmental benefit to switching to uv (by default).

**Comment by humitos** [comment]
We had a call yesterday and we decided to start thinking in a way to integrate `uv` natively into our config file following the patterns we have already established for `pip`: https://docs.readthedocs.com/platform/stable/config-file/v2.html#python

I'm re-opening this issue so we can prioritize it and put it in our roadmap. We appreciate any feedback you may have about _How would you like to declare that you are using `uv` in your `readthedocs.yaml` file?_. That will help us to shape this feature in a way that users already expect 👍🏼

**Comment by ddelange** [comment]
> How would you like to declare that you are using `uv` in your `readthedocs.yaml` file?

this would be my preferred approach for the simple `uv pip install` case (solving for speedup purely): https://github.com/readthedocs/readthedocs.org/issues/11289#issuecomment-3694653123
but there are probably also arguments for a deeper uv ecosystem integration (`uv sync` etc)

**Comment by cjw296** [comment]
@humitos - ideally it'd be super simple, couple of cases I can think of

Simplest:
```yaml
python:
  install:
  - method: uv
```

Custom args, couple of examples:
```yaml
python:
  install:
  - method: uv
  - options: "--all-groups --all-extras"
```

```yaml
python:
  install:
  - method: uv
  - options: "--group docs"
```

For repos where the pyproject isn't at the root:

```yaml
python:
  install:
  - method: uv
  - path: some/folders
  - options: "--group docs"
```

**Comment by effigies** [comment]
I would probably make `uv` vs `pip+virtualenv` (or `venv`?) a top-level option, e.g., your current example could be switched over to `uv pip install` with:

```diff
 python:
+  installer: uv  # or pip
   install:
     - requirements: docs/requirements.txt
     - method: pip
       path: .
       extra_requirements:
         - docs
     - method: pip
       path: another/package
```

Then for `uv sync`, you would write:

```yaml
python:
  installer: uv
  install:
    - method: sync
      # defaults below
      path: .  # UV_WORKING_DIR
      extras: []  # --extra
      groups: []  # --group
      frozen: true  # --frozen
      active: true  # --active
      options: ""
```

**Comment by janbridley** [comment]
> [@humitos](https://github.com/humitos) - ideally it'd be super simple, couple of cases I can think of
> 
> Simplest:
> 
> python:
>   install:
>   - method: uv
> Custom args, couple of examples:
> 
> python:
>   install:
>   - method: uv
>   - options: "--all-groups --all-extras"
>  python:
>   install:
>   - method: uv
>   - options: "--group docs"
>
> ...

As mentioned by @cjw296, I think this approach is pretty much optimal. It's a one-line swap for most users, with the following being standard for most of my org's projects:

```diff
python:
  install:
-    - method: pip
+    - method: uv
      path: .
      extra_requirements:
        - doc
```

**Comment by agjohnson** [comment]
I don't feel we should try to overload the `method: pip` configuration, there are configurations options there that don't make sense for `uv` and options for `uv` that won't make sense for `pip`.

I'd also prefer to promote `uv sync` style usage over requirements files -- requirements files are an old relic from the world outside Python packaging. Python packaging has had modern alternatives to these for some time now, we could use to promote these patterns more.

We can still provide `uv pip` operations for compatibility, but I'd mostly want to frame uv support around proper lock file usage.

The extra configuration option noted above are good additions, but as a base I think we should aim for:

```yaml
python:
  install:
    - method: uv
      operation: sync
      # or
      operation: pip
      ...
    # Or explicit different method for pip
    - method: uv pip
    # Or something similar to distinguish the sub command used
```

**Comment by henryiii** [comment]
If you did that, you'd also want to be able to select groups; documentation might not be in the `dev` group.

**Comment by humitos** [comment]
Just a heads up here.

We are moving forward with the implementation of `uv` support. Soon, you will be able to do to manage the common use cases:

```yaml
python:
  install:
    - method: uv
      command: sync
```

or

```yaml
python:
  install:
    - method: uv
      command: pip
      requirements: docs/requirements.txt
```

- The design doc is at https://github.com/readthedocs/readthedocs.org/pull/12868
- The implementation's PR at: https://github.com/readthedocs/readthedocs.org/pull/12879
